kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PulsarTopicProducerStateManagerSnapshotBuffer.java

```diff
@@ -53,28 +57,70 @@ public class PulsarTopicProducerStateManagerSnapshotBuffer implements ProducerSt
 
     private synchronized CompletableFuture<Reader<ByteBuffer>> ensureReaderHandle() {
         if (reader == null) {
-            reader = pulsarClient.newReaderBuilder()
+            CompletableFuture<Reader<ByteBuffer>> newReader = pulsarClient.newReaderBuilder()
                     .topic(topic)
                     .startMessageId(MessageId.earliest)
                     .readCompacted(true)
                     .createAsync();
+            reader = newReader;
+
+            newReader.whenComplete((r, error) -> {
+                if (error != null) {
+                    discardReader(newReader);
+                }
+            });
         }
         return reader;
     }
 
+    private synchronized void discardReader(CompletableFuture<Reader<ByteBuffer>> oldReader) {
+        if (reader == oldReader) {
+            reader = null;
+        }
+    }
+
+    private synchronized void discardReader(Reader<ByteBuffer> oldReader) {
+        if (reader.isDone() && !reader.isCompletedExceptionally()) {
+            Reader<ByteBuffer> newReader = null;
+            try {
+                newReader = reader.get(0, TimeUnit.MILLISECONDS);
+                if (newReader == oldReader) {
+                    reader = null;
+                }
+            } catch (Exception exception) {
+                log.warn("Failed to get reader handle for topic {}, discard the reader.", topic, exception);
+                reader = null;
+            }
+        }
```

Comment:
The original code is this, but there is a spotbugs problem(refer to [this](https://github.com/spotbugs/spotbugs/issues/1001). /cc @Demogorgon314 @BewareMyPower 
```
if (reader.isDone()
        && !reader.isCompletedExceptionally()
        && reader.getNow(null) == oldReader) {
    reader = null;
}
```